mainnet-contracts/src/PufferProtocolLogic.sol

```diff
@@ -0,0 +1,667 @@
+            }),
+            deprecated_burntVTs: 0
+        });
+
+        // The bond is converted to pufETH at the current exchange rate
+        uint256 pufETHBondAmount = _PUFFER_VAULT.depositETH{ value: bondAmountEth }(address(this));
+
+        uint256 pufferModuleIndex = $.pendingValidatorIndices[moduleName];
+
+        // No need for SafeCast
+        $.validators[moduleName][pufferModuleIndex] = Validator({
+            pubKey: data.blsPubKey,
+            status: Status.PENDING,
+            module: address($.modules[moduleName]),
+            bond: uint96(pufETHBondAmount),
+            node: msg.sender,
+            numBatches: numBatches
+        });
+
+        // Increment indices for this module and number of validators registered
+        unchecked {
+            $.nodeOperatorInfo[msg.sender].epochPrice = epochCurrentPrice;
+            $.nodeOperatorInfo[msg.sender].validationTime += (msg.value - bondAmountEth);
+            ++$.nodeOperatorInfo[msg.sender].pendingValidatorCount;
+            ++$.pendingValidatorIndices[moduleName];
+            ++$.moduleLimits[moduleName].numberOfRegisteredValidators;
+        }
+
+        emit IPufferProtocol.NumberOfRegisteredValidatorsChanged({
+            moduleName: moduleName,
+            newNumberOfRegisteredValidators: $.moduleLimits[moduleName].numberOfRegisteredValidators
+        });
+        emit IPufferProtocol.ValidatorKeyRegistered({
+            pubKey: data.blsPubKey,
+            pufferModuleIndex: pufferModuleIndex,
+            moduleName: moduleName,
+            numBatches: numBatches
+        });
+    }
+
+    /**
+     * @dev This function should only be called by the PufferProtocol contract through a delegatecall
+     */
+    function _requestConsolidation(bytes32 moduleName, uint256[] calldata srcIndices, uint256[] calldata targetIndices)
+        external
+        payable
+        override
+    {
+        if (srcIndices.length == 0) {
+            revert InputArrayLengthZero();
+        }
+        if (srcIndices.length != targetIndices.length) {
+            revert InputArrayLengthMismatch();
+        }
+
+        ProtocolStorage storage $ = _getPufferProtocolStorage();
+
+        bytes[] memory srcPubkeys = new bytes[](srcIndices.length);
+        bytes[] memory targetPubkeys = new bytes[](targetIndices.length);
+        Validator storage validatorSrc;
+        Validator storage validatorTarget;
+        for (uint256 i = 0; i < srcPubkeys.length; i++) {
+            require(srcIndices[i] != targetIndices[i], InvalidValidator());
+            validatorSrc = $.validators[moduleName][srcIndices[i]];
+            require(validatorSrc.node == msg.sender && validatorSrc.status == Status.ACTIVE, InvalidValidator());
+            srcPubkeys[i] = validatorSrc.pubKey;
+            validatorTarget = $.validators[moduleName][targetIndices[i]];
+            require(validatorTarget.node == msg.sender && validatorTarget.status == Status.ACTIVE, InvalidValidator());
```

Comment:
Did we doument anywhere that scenario, where Source consolidates into a bad target and NoOp loses money?